pkg/spi-shared/tokenstorage/vault.go

```diff
@@ -0,0 +1,137 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tokenstorage
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	k8s "sigs.k8s.io/controller-runtime/pkg/client"
+
+	vault "github.com/hashicorp/vault/api"
+	auth "github.com/hashicorp/vault/api/auth/kubernetes"
+	api "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const vaultDataPathFormat = "spi/data/%s/%s"
+
+type vaultTokenStorage struct {
+	*vault.Client
+}
+
+// NewVaultStorage creates a new `TokenStorage` instance using the provided Vault instance.
+func NewVaultStorage(k8sClient k8s.Client, role string, vaultHost string, serviceAccountToken string) (TokenStorage, error) {
+	config := vault.DefaultConfig()
+	config.Address = vaultHost
+	vaultClient, err := vault.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	var k8sAuth *auth.KubernetesAuth
+	if serviceAccountToken == "" {
+		k8sAuth, err = auth.NewKubernetesAuth(role)
+	} else {
+		k8sAuth, err = auth.NewKubernetesAuth(role, auth.WithServiceAccountTokenPath(serviceAccountToken))
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	authInfo, err := vaultClient.Auth().Login(context.TODO(), k8sAuth)
+	if err != nil {
+		return nil, err
+	}
+	if authInfo == nil {
+		return nil, fmt.Errorf("no auth info was returned after login to vault")
+	}
+	return &NotifyingTokenStorage{
+		Client:       k8sClient,
+		TokenStorage: &vaultTokenStorage{vaultClient},
+	}, nil
+}
+
+func (v *vaultTokenStorage) Store(ctx context.Context, owner *api.SPIAccessToken, token *api.Token) error {
+	data := map[string]interface{}{
+		"data": token,
+	}
+	path := getVaultPath(owner)
+	s, err := v.Client.Logical().Write(path, data)
+	if err != nil {
+		return err
+	}
+	if s == nil {
+		return fmt.Errorf("failed to store the token, no error but returned nil")
+	}
+	for _, w := range s.Warnings {
+		logf.FromContext(ctx).Info(w)
+	}
+
+	return nil
+}
+
+func (v *vaultTokenStorage) Get(ctx context.Context, owner *api.SPIAccessToken) (*api.Token, error) {
+	path := getVaultPath(owner)
+
+	secret, err := v.Client.Logical().Read(path)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil || secret.Data == nil || len(secret.Data) == 0 || secret.Data["data"] == nil {
+		logf.FromContext(ctx).Info("no data found in vault at", "path", path)
+		return nil, nil
+	}
+	for _, w := range secret.Warnings {
+		logf.FromContext(ctx).Info(w)
+	}
+	data, dataOk := secret.Data["data"]
+	if !dataOk {
+		return nil, fmt.Errorf("corrupted data in Vault at '%s'", path)
+	}
+
+	return parseToken(ctx, data)
+}
+
+func parseToken(ctx context.Context, data interface{}) (*api.Token, error) {
+	// We only get `interface{}` from Vault. In order to get `Token{}` struct,
+	// we first marshall interface into JSON, and then unmarshall it back to `Token{}` struct.
+	jsonData, err := json.Marshal(data)
```

Comment:
what actual type is the `data` at runtime? it feels weird to convert it to JSON and then right back into a struct.

Reply:
it's really `interface{}`. comes from this map https://github.com/hashicorp/vault/blob/main/api/secret.go#L26

Reply:
if you have better idea how to convert it to our `Token`, I'm all in

Reply:
well, the json marshaller is able to make sense of it and extract the data from it, so it cannot be just `interface{}` at runtime. I was just advocating to see if this cannot be reliably cast into at least `map[string]interface{}` or something from which we could extract the data, too, without the whole serde dance...

Reply:
I've changed it to mapstructure decoder (comes with vault dependency), but I'm not sure about it. I can also get the fields individually, but it's not very nice code. One benefit I see with using json marshaller is, that it will follow Token struct tags. But I'm not sure it's worth as I agree it's kind of nasty hack.

WDYT?

Reply:
I've tried to debug into the mapstructure and it is the same reflection-ridden madness as json decoder :)

what about just this (tested only with the test vault cluster, not the real one) (`json.Number` is just a typedef of `string` so the conversions in the below code are noops and the parsing into an actual `uint64` is inevitable):

I know that this code breaks as soon as we change the format of the data stored in Vault but that would most possibly be true for serialization solutions, too, if the data in vault is in the older format. At the same time this is going to be called a lot, so I think we could afford to be specific.

```go
func parseToken(data interface{}) (*api.Token, error) {
	token := &api.Token{}
	dataMap, ok := data.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected data")
	}

	token.AccessToken = nonNilOrDefault(dataMap["access_token"], "").(string)
	token.TokenType = nonNilOrDefault(dataMap["token_type"], "").(string)
	token.RefreshToken = nonNilOrDefault(dataMap["refresh_token"], "").(string)
	expiryNum := nonNilOrDefault(dataMap["expiry"], json.Number("0")).(json.Number)

	expiry, err := strconv.ParseUint(expiryNum.String(), 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid expiry value: %v", expiryNum)
	}
	token.Expiry = expiry

	return token, nil
}

func nonNilOrDefault(val interface{}, defaultValue interface{}) interface{} {
	if val == nil {
		val = defaultValue
	}

	return val
}
```

Reply:
ok, next try :) It's basically your suggestion, except I very much don't like `interface{}`s in function args/return (doesn't lead to anything good from my experience), so I've wrote 2 separate functions for `string` and `uint64` (we can refactor to generics once we switch to 1.18 🚀). If you're ok with it, I'll write few tests.

Btw, can you please explain more why you don't like generic solution? I mean JSON or mapstructure one? Is it because it's hacky? Performance? Prone to errors?

Reply:
For me it's performance.

Reply:
ok. I agree with performance, as it is not using reflection (debatable how it is performance critical place though). I wouldn't be so sure with low maintenance cost. We need to always think to update this function when updating `Token`, which might lead to unnecessary debug sessions. Also, when we would like to enforce properties in `Token` json tags, the logic becomes tricky. But we can always refactor later.

Reply:
We also need to think about how to deal with existing data when updating the `Token`. At that point I think the manual solution might even offer more flexibility. But I actually am not too worried about this. I don't want to block the PR because of this.